### PR TITLE
Feature Development: Merging Dev Changes into Main

### PR DESCRIPTION
**Pull request by RaghadAbusamor**

### CODE REVIEW

File: Enums/UserType.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿
+
```
Comment (hayatzsala) [inline Enums/UserType.cs]: empty lines should removed

File: Enums/Filterparameter.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (hayatzsala) [inline Enums/Filterparameter.cs]: unused libs

File: FileSystem/FileOperations.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    DateTime departureDate = DateTime.Parse(values[3]); // Assuming date is in a valid format
+                    string departureAirport = values[4];
+                    string arrivalAirport = values[5];
+                    FlightClass flightClass = Enum.Parse<FlightClass>(values[6]); // Assuming flight class is in correct format
+                    decimal price = decimal.Parse(values[7]); // Assuming price is in correct format
+
+                    // Create a new Flight object and add it to the list
+                    Flight flight = new Flight(flightNumber, departureCountry, destinationCountry, departureDate, departureAirport, arrivalAirport, flightClass, price);
+                    flights.Add(flight);
+                }
```
Comment (hayatzsala) [inline FileSystem/FileOperations.cs]: @RaghadAbusamor Good job, But search for CsvHelper library and try to use it. As a general tip, whenever we have a library that can help us from write code from scratch so let's use it.

File: FileSystem/FileOperations.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            return flights;
+        }
+
+        public static async Task WriteToCSVAsync(string filePath, List<string[]> data)
+        {
+            using (var writer = new StreamWriter(filePath))
+            {
+                foreach (var row in data)
+                {
+                    await writer.WriteLineAsync(string.Join(",", row));
```
Comment (hayatzsala) [inline FileSystem/FileOperations.cs]: same here, try to using csvHelper

File: UserRole/User.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+            Id = id;
+        }
+    }
+}
+
+
+
+
+
+
```
Comment (hayatzsala) [inline UserRole/User.cs]: empty lines

File: UserRole/UserTypeSelection.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
+﻿using AirportTicketBookingSystem.Enums;
+using AirportTicketBookingSystem.Flights;
+
+namespace AirportTicketBookingSystem.UserRole
+{
+    public static class UserTypeSelection
```
Comment (hayatzsala) [inline UserRole/UserTypeSelection.cs]: class name should be more descriptive, for example: UserTypeSelectionService

File: UserRole/UserTypeSelection.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                while (!int.TryParse(Console.ReadLine(), out id))
+                {
+                    Console.WriteLine("Invalid input. Please enter a valid ID.");
+                    Console.Write("Enter your ID: ");
+                }
+                // Create a new manager object with the entered details
+                var manager = new Manager(name, id);
+                // Display manager operations
+                ManagerOperations managerOperation = new ManagerOperations();
+                managerOperation.SelectManagerOperation(manager);
```
Comment (hayatzsala) [inline UserRole/UserTypeSelection.cs]: this code block the best practice of it to extract into private method and here just call it

File: UserRole/UserTypeSelection.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                string passportNumber = Console.ReadLine();
+
+                Console.Write("Enter your Phone Number: ");
+                string PhoneNumber = Console.ReadLine();
+                // Create a new passenger object with the entered details
+                var passenger = new Passenger(name, id, PhoneNumber, passportNumber);
+                PassengerOperations passengerOperations = new PassengerOperations();
+                // Display the passenger operations
+                passengerOperations.SelectPassengerOperation(passenger);
+            }
```
Comment (hayatzsala) [inline UserRole/UserTypeSelection.cs]: same here, extract to method

File: UserRole/UserTypeSelection.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                Console.Write("Enter your ID: ");
+                int id;
+                while (!int.TryParse(Console.ReadLine(), out id))
+                {
+                    Console.WriteLine("Invalid input. Please enter a valid ID.");
+                    Console.Write("Enter your ID: ");
+                }
+                // Create a new manager object with the entered details
+                var manager = new Manager(name, id);
+                // Display manager operations
```
Comment (hayatzsala) [inline UserRole/UserTypeSelection.cs]: these steps are not need actually, first thing: the requirement doesn't ask to create/have users, the second thing: why we need to enter the name and the id and there is not authentication process so it's out of scoop

File: Flight/FlightManagement.cs
Comment (hayatzsala) [inline Flight/FlightManagement.cs:1]: not a best practice to add the model and it's functionality in same class & file, so please extract the functionality into other class with name FlightManagmentService for example

File: Flight/ManagerOperations.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                Console.Clear();
+            }
+        }
+
+        private static void ManagerMenu()
+        {
+            Console.WriteLine("Manager Operations Menu:");
+            Console.WriteLine($"1. {ManagerOperation.FilterBookings}");
+            Console.WriteLine($"2. {ManagerOperation.BatchFlightUpload}");
+            Console.WriteLine($"3. Exit");
```
Comment (hayatzsala) [inline Flight/ManagerOperations.cs]: use the enum here as other option, why Exit is static value here?

File: Flight/PassengerOperations.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        private static void PasengerMenu()
+        {
+            Console.WriteLine("Passenger Operations Menu:");
+            Console.WriteLine($"1. {PassengerOperation.BookFlight}");
+            Console.WriteLine($"2. {PassengerOperation.ListAllFlights}");
+            Console.WriteLine($"3. {PassengerOperation.CancelBooking}");
+            Console.WriteLine($"4. {PassengerOperation.EditBooking}");
+            Console.WriteLine($"5. {PassengerOperation.ViewPersonalBookings}");
+            Console.WriteLine($"6. {PassengerOperation.SearchBooking}");
+            Console.WriteLine($"7. Exit");
```
Comment (hayatzsala) [inline Flight/PassengerOperations.cs]: same here use the enum option instead of static value

File: Flight/Flight.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+using AirportTicketBookingSystem.ModelValidation;
+
+namespace AirportTicketBookingSystem.Flights
+{
+    public class Flight
+    {
+
+        public Flight(string flightNumber, string departureCountry, string destinationCountry,
+                      DateTime departureDate, string departureAirport, string arrivalAirport,
+                      FlightClass flightClass, decimal price)
```
Comment (hayatzsala) [inline Flight/Flight.cs]: usually in any method, in clean code, if the parameter is more than 4 then pass an object. so it's un-clean to have constructor with parameter of all properties

File: Flight/Flight.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        [DynamicValidation("DateTime", true, "", ErrorMessage = "Departure date is required")]
+        [FutureDate(ErrorMessage = "Departure date must be today or in the future")]
+        public DateTime DepartureDate { get; set; }
+
+        [DynamicValidation("string", true, "", ErrorMessage = " Destination Country is required   ")]
+        public string DestinationCountry { get; set; }
+
+        [DynamicValidation("string", true, "", ErrorMessage = "Flight number is required")]
+        public string FlightNumber { get; set; }
+        [DynamicValidation("decimal", true, "", ErrorMessage = "Price is required")]
```
Comment (hayatzsala) [inline Flight/Flight.cs]: in clean code, we should follow vertical space density which is have empty line between unrelated code to be easy to read, so leave line between these

File: Flight/FlightManagement.cs [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+        {
+            Flights = new List<Flight>();
+            PassengerFlights = new List<Flight>();
+            AddDefaultFlight();
+        }
+
+        public List<Flight> Flights { get; set; }
+        public List<Flight> PassengerFlights { get; set; }
+
+        public async Task BatchFlightUpload()
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: when the method is async so always add async suffix

File: Flight/FlightManagement.cs [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+            var selectedFlight = Flights.FirstOrDefault(f => f.FlightNumber.Equals(flightNumber, StringComparison.OrdinalIgnoreCase));
+            if (selectedFlight != null)
+            {
+                PassengerFlights.Add(selectedFlight);
+                Console.WriteLine($"Flight {flightNumber} added to your bookings.");
+            }
+            else
+            {
+                Console.WriteLine($"Flight with number {flightNumber} not found.");
+            }
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: invert the if condition here, and add return statement if the flight number not found it will reduce the Cyclomatic complexity.
Read https://bytedev.medium.com/cyclomatic-complexity-c-net-5d5d14cf7fe9

File: Flight/FlightManagement.cs [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+            var selectedFlight = PassengerFlights.FirstOrDefault(f => f.FlightNumber.Equals(flightNumber, StringComparison.OrdinalIgnoreCase));
+            if (selectedFlight != null)
+            {
+                PassengerFlights.Remove(selectedFlight);
+                Console.WriteLine($"Flight {flightNumber} Canceled.");
+            }
+            else
+            {
+                Console.WriteLine($"Flight with number {flightNumber} not found.");
+            }
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: same comment here, invert the condition

File: CSVFiles/FlightSheet.csv [hunk 17]
```diff
@@ -0,0 +1,2 @@
+FlightNumber,DepartureCountry,DestinationCountry,DepartureDate,DepartureAirport,ArrivalAirport,Class,Price
+111,palestine,jordan,5/ 4/ 2025,palestine ,QueenAlia,FirstClass,2000
```
Comment (hayatzsala) [inline CSVFiles/FlightSheet.csv]: any flight have 3 classes and each class has own price so you should handle for each flight the 3 classes with each price

File: Flight/FlightManagement.cs (hunk 14 above)
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: It looks like you don't implement the exact requirement from this method, the Task needs to depend on the file system as storage layer, so the Idea is we need a file with all flights, and they should be loaded into Flights, so if I'm a manger I need to add batch/bulk of flights instead of add one by one I will get a new flight file and upload so add the new flights from my file to the origin file, then the Flights list also.

File: Flight/FlightManagement.cs [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+            {
+                Console.WriteLine($"Flight management error: {ex.Message}");
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"An unexpected error occurred: {ex.Message}");
+            }
+        }
+
+        public void BookFlight()
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: add the booked flight in a dynamic property for the object initiated is something not match the requirements in the Task, you should have a file for booked flights with flight number and passenger Id and other details for passenger booking

File: Flight/FlightManagement.cs [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+                PassengerFlights.Add(selectedFlight);
+                Console.WriteLine($"Flight {flightNumber} added to your bookings.");
+            }
+            else
+            {
+                Console.WriteLine($"Flight with number {flightNumber} not found.");
+            }
+        }
+
+        public void CancelBooking()
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: Then, this method should remove the record from the file

File: Flight/FlightManagement.cs [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+                        Console.WriteLine("Invalid class value.");
+                        return;
+                    }
+                    break;
+                default:
+                    Console.WriteLine("Invalid parameter.");
+                    return;
+            }
+
+            List<Flight> filteredFlights = Flights.Where(filterPredicate).ToList();
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: good use for delegate here

File: Flight/FlightManagement.cs [hunk 21]
```diff
@@ -0,0 +1,275 @@
...
+            {
+                Console.WriteLine($"An error occurred: {ex.Message}");
+            }
+        }
+
+        private void AddDefaultFlight()
+        {
+            Flights.Add(new Flight("ABC123", "USA", "UK", DateTime.Now.AddDays(1), "JFK", "LHR", FlightClass.Business, 1000));
+            Flights.Add(new Flight("RST", "USA", "UK", DateTime.Now.AddDays(1), "JFK", "LHR", FlightClass.Business, 2000));
+        }
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: this method should be used, since we should upload the flights from file

File: Flight/FlightManagement.cs [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+                    Console.WriteLine("Thank you for using our flight booking service.");
+                }
+            }
+            catch (FlightManagementException ex)
+            {
+                Console.WriteLine($"An error occurred: {ex.Message}");
+            }
+        }
+
+        public void ViewPersonalBookings()
```
Comment (hayatzsala) [inline Flight/FlightManagement.cs]: the idea of this method to passing the user/passenger ID that I used at the beginning and filter the booked flights depend on the passengerId

File: Flight/PassengerOperations.cs [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+
+                    case PassengerOperation.EditBooking:
+                        flightManagement.EditBooking();
+                        break;
+
+                    case PassengerOperation.ViewPersonalBookings:
+                        flightManagement.ViewPersonalBookings();
+                        break;
+
+                    case PassengerOperation.SearchBooking:
```
Comment (hayatzsala) [inline Flight/PassengerOperations.cs]: here you used the filtering method but it shouldn't, there is a big difference between Search & Filter in the Business layer, 
Search: I have a complex criterion with different properties that should be all filled and all matched on the criteria properties, all property should be filled.
Filter: I have single criteria property so I will pass one property to get all matched, one parameter is enough

**Review by hayatzsala — CHANGES_REQUESTED** [review]
@RaghadAbusamor The task is not structured as expected in the requirements, so Please read the requirements again, accurately understand the functionality, then check the comments and fix/re-design fo

File: FileSystem/FileOperations.cs [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+        using (var reader = new StreamReader(filePath))
+        using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
+        {
+            // Skip header record
+            await csv.ReadAsync();
+            csv.ReadHeader();
+
+            while (await csv.ReadAsync())
+            {
+                Flight flight = csv.GetRecord<Flight>();
```
Comment (hayatzsala) [inline FileSystem/FileOperations.cs]: try GetRecords instead of while loop

File: FileSystem/FileOperations.cs [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public static async Task<List<Flight>> ReadFlightsFromCSVAsync(string filePath)
+    {
+        List<Flight> flights = new List<Flight>();
+
+        using (var reader = new StreamReader(filePath))
+        using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
+        {
+            // Skip header record
+            await csv.ReadAsync();
```
Comment (hayatzsala) [inline FileSystem/FileOperations.cs]: no need now to skip the header

File: Flight/FiilterBooking.cs
Comment (hayatzsala) [inline Flight/FiilterBooking.cs:1]: why this class?

**Review by hayatzsala — CHANGES_REQUESTED** [review]

File: Flights/FlightManagement.cs
Comment (hayatzsala) [inline Flights/FlightManagement.cs:1]: what is the different between FlightManagment & FlightManagmentService ??? Please clean up the code in 2 files

File: CSVFiles/FlightSheet.csv (hunk 17 above)
Comment (hayatzsala) [reply CSVFiles/FlightSheet.csv]: Again! this comment from requirement and didn't addressed

File: FileSystem/FileOperations.cs (hunk 24 above)
Comment (hayatzsala) [reply FileSystem/FileOperations.cs]: check this as well

**Review by hayatzsala — CHANGES_REQUESTED** [review]
Please check all the comments

**Comment by hayatzsala** [comment]
Where is the .sln & .csproj files? this repo we couldn't run the code?

File: Flights/FlightManagement.cs
Comment (RaghadAbusamor) [reply Flights/FlightManagement.cs:1]: FlightManagement oversees the overall flight management process, while FlightManagementService handles the implementation details of specific flight management tasks like displayflight and FindBookingByFlightNumber.

**Review by hayatzsala — COMMENTED** [review]
please add .gitignore and make the MR clean

**Review by hayatzsala — APPROVED** [review]
